### PR TITLE
geometry:optimization: AddPointInNonnegativeScalingConstraints

### PR DESCRIPTION
**Pull request by RussTedrake**
This is another well-defined operation on convex sets, and it is the workhorse for the MIP formulation of the shortest-path problem.

---
This change is [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/15286)

### CODE REVIEW

**Review by RussTedrake — COMMENTED** [review]
+@avalenzu for feature review, please.
Note: this PR contains 2 commits.  The first is from #15285, which is only a refactoring into separate files.  Please ignore that here, and only review the second commit.  (I've set reviewable to "review each commit separately" mode).

> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/15286#-:-MdWf1HJ50UjGDOU6c9Z:bc6sjsr)* status: LGTM missing from assignee avalenzu, needs platform reviewer assigned, needs at least two assigned reviewers, commits need curation (https://drake.mit.edu/reviewable.html#curated-commits) (waiting on @avalenzu)

**Comment by RussTedrake** [comment]
cc @tobiamarcucci , in case you are able to look.  (only check the difference between r1 and r2)

File: geometry/optimization/convex_set.h [hunk 1]
```diff
@@ -87,20 +89,40 @@ class ConvexSet : public ShapeReifier {
...
   }
 
+  /** Adds the convex constraints to imply
+  @verbatim
+  x ∈ t S,
+  t ≥ 0,
+  @endverbatim
+  where x is a point in ℜⁿ (with n the ambient_dimension), S is this convex set,
+  and t is a scalar.  Note that many applications of this method also require
+  the S is bounded (in addition to being convex); e.g. in order to ensure that
```
Comment (TobiaMarcucci) [inline geometry/optimization/convex_set.h:103]: the S is bounded -> S to be bounded?

File: geometry/optimization/convex_set.h [hunk 2]
```diff
@@ -87,20 +89,40 @@ class ConvexSet : public ShapeReifier {
   void AddPointInSetConstraint(
       solvers::MathematicalProgram* prog,
       const Eigen::Ref<const solvers::VectorXDecisionVariable>& vars) const {
+    DRAKE_DEMAND(vars.size() == ambient_dimension());
     return DoAddPointInSetConstraint(prog, vars);
   }
 
+  /** Adds the convex constraints to imply
+  @verbatim
+  x ∈ t S,
```
Comment (TobiaMarcucci) [inline geometry/optimization/convex_set.h:98]: One technical observation, probably my bad, sorry for this. It is not exactly true that this method enforces `x ∈ t S`, because that would imply that `x = 0` if `t = 0`, even if `S` is unbounded.

To make this very precise I see two options:
- Explicitly require the set `S` to be bounded for this method to be used.
- Or say that this enforces the constraint `x ∈ t S + rec(S)` where `rec(S)` is the recession cone of `S` (the asymptotic directions in which `S` is not bounded). In fact, when `t = 0` the perspective is actually enforcing `x ∈ rec(S)` (which is `x ∈ {0}` for a bounded set), while when `t > 0` it enforces `x ∈ t S + rec(S) = t S`.

**Review by RussTedrake — COMMENTED** [review]
-@avalenzu 
+@hongkai-dai for feature review, please

> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/15286#-:-MdXCVmx8lZZ-z-65Y4f:bvlumiy)* status: 2 unresolved discussions, LGTM missing from assignee hongkai-dai, needs platform reviewer assigned, needs at least two assigned reviewers, commits need curation (https://drake.mit.edu/reviewable.html#curated-commits) (waiting on @hongkai-dai and @TobiaMarcucci)

___
*[geometry/optimization/convex_set.h, line 98 at r2](https://reviewable.io/reviews/robotlocomotion/drake/15286#-MdWeujHyImNgBIqqbD6-r2-98:-MdXCBeqE8yYnEMsf3dm:bz2hk4z) ([raw file](https://github.com/robotlocomotion/drake/blob/c706fa2466b34207d28d96270fe235d27b7dde80/geometry/optimization/convex_set.h#L98)):*
<details><summary><i>Previously, TobiaMarcucci (Tobia Marcucci) wrote…</i></summary><blockquote>

One technical observation, probably my bad, sorry for this. It is not exactly true that this method enforces `x ∈ t S`, because that would imply that `x = 0` if `t = 0`, even if `S` is unbounded.

To make this very precise I see two options:
- Explicitly require the set `S` to be bounded for this method to be used.
- Or say that this enforces the constraint `x ∈ t S + rec(S)` where `rec(S)` is the recession cone of `S` (the asymptotic directions in which `S` is not bounded). In fact, when `t = 0` the perspective is actually enforcing `x ∈ rec(S)` (which is `x ∈ {0}` for a bounded set), while when `t > 0` it enforces `x ∈ t S + rec(S) = t S`.
</blockquote></details>

Done. Thanks for catching it.  I've taken another pass.
___
*[geometry/optimization/convex_set.h, line 103 at r2](https://reviewable.io/reviews/robotlocomotion/drake/15286#-MdWeujHyImNgBIqqbD6-r2-103:-MdXCMnm0_sfH7YB0xjF:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/c706fa2466b34207d28d96270fe235d27b7dde80/geometry/optimization/convex_set.h#L103)):*
<details><summary><i>Previously, TobiaMarcucci (Tobia Marcucci) wrote…</i></summary><blockquote>

the S is bounded -> S to be bounded?
</blockquote></details>

Done.

**Review by RussTedrake — COMMENTED** [review]
The earlier PR has landed, so I've rebased;  this one is back down to a single commit.

> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/15286#-:-MdXS00l93AJPBM-N5XW:bl65zg)* status: 2 unresolved discussions, LGTM missing from assignee hongkai-dai, needs platform reviewer assigned, needs at least two assigned reviewers (waiting on @hongkai-dai and @TobiaMarcucci)

**Review by hongkai-dai — COMMENTED** [review]
> Reviewed 7 of 8 files at r2, 1 of 1 files at r3, 5 of 5 files at r4.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/15286)* status: 7 unresolved discussions, LGTM missing from assignee hongkai-dai, needs platform reviewer assigned, needs at least two assigned reviewers (waiting on @RussTedrake and @TobiaMarcucci)

___
*[geometry/optimization/convex_set.h, line 98 at r2](https://reviewable.io/reviews/robotlocomotion/drake/15286#-MdWeujHyImNgBIqqbD6-r2-98:-MdXv3Tu1N7iEw0OkKZ6:bhlyc13) ([raw file](https://github.com/robotlocomotion/drake/blob/c706fa2466b34207d28d96270fe235d27b7dde80/geometry/optimization/convex_set.h#L98)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary><blockquote>

Done. Thanks for catching it.  I've taken another pass.
</blockquote></details>

BTW, when writing `tS + rec(S)`, would it be more clear to use Minkowski sum ⊕ to denote it is the summation of sets?
___
*[geometry/optimization/convex_set.h, line 102 at r4](https://reviewable.io/reviews/robotlocomotion/drake/15286#-MdXvKAnAp5eUxpU6ePr:-MdXvKAnAp5eUxpU6ePs:b-ccc1tz) ([raw file](https://github.com/robotlocomotion/drake/blob/ae763485c1d12931f2dc529e6a253c475e556f9e/geometry/optimization/convex_set.h#L102)):*
> ```objc
>   @endverbatim
>   where x is a point in ℜⁿ (with n the ambient_dimension), S is this convex set,
>   S is bounded, and t is a scalar.
> ```

I thought we don't require `S is bounded`?
___
*[geometry/optimization/hyperellipsoid.cc, line 67 at r4](https://reviewable.io/reviews/robotlocomotion/drake/15286#-MdXwUmK3fdRTPKYCMN4:-MdXwUmK3fdRTPKYCMN5:b-jtdgfd) ([raw file](https://github.com/robotlocomotion/drake/blob/ae763485c1d12931f2dc529e6a253c475e556f9e/geometry/optimization/hyperellipsoid.cc#L67)):*
> ```cpp
>     const Eigen::Ref<const VectorXDecisionVariable>& vars) const {
>   // 1.0 ≥ |A * (vars - center)|_2, written as
>   // z₀ ≥ |z₁...ₘ|₂ with z = A_cone* x + b_cone.
> ```

A_cone * vars + b_cone? x is not mentioned before.
___
*[geometry/optimization/hyperellipsoid.cc, line 84 at r4](https://reviewable.io/reviews/robotlocomotion/drake/15286#-MdXy1iy9wZ9M_Pr6-5Q:-MdXy1iy9wZ9M_Pr6-5R:byshurk) ([raw file](https://github.com/robotlocomotion/drake/blob/ae763485c1d12931f2dc529e6a253c475e556f9e/geometry/optimization/hyperellipsoid.cc#L84)):*
> ```cpp
>   std::vector<Binding<Constraint>> constraints;
>   // t ≥ |A * (x - t*center)|_2, written as
>   // z₀ ≥ |z₁...ₘ|₂ with z = A_cone* x.
> ```

z = A_cone * [x;t]
___
*[geometry/optimization/hyperellipsoid.cc, line 90 at r4](https://reviewable.io/reviews/robotlocomotion/drake/15286#-MdXyAghAMQkk_gvtKs7:-MdXyAghAMQkk_gvtKs8:b-pa45yv) ([raw file](https://github.com/robotlocomotion/drake/blob/ae763485c1d12931f2dc529e6a253c475e556f9e/geometry/optimization/hyperellipsoid.cc#L90)):*
> ```cpp
>   A_cone(0, n) = 1.0;  // z₀ = t.
>   A_cone.block(1, 0, m, n) = A_;
>   A_cone.block(1, n, m, 1) = center_;
> ```

This should be `-A * center_` instead of `center_`? And maybe we need a unit test with non-identity `A` and non-zero `center_` to catch it?
___
*[geometry/optimization/test/hpolyhedron_test.cc, line 210 at r4](https://reviewable.io/reviews/robotlocomotion/drake/15286#-MdXyc3Y1omaz8A35Kc2:-MdXyc3Y1omaz8A35Kc3:b-xd6wcl) ([raw file](https://github.com/robotlocomotion/drake/blob/ae763485c1d12931f2dc529e6a253c475e556f9e/geometry/optimization/test/hpolyhedron_test.cc#L210)):*
> ```cpp
> 
> GTEST_TEST(HPolyhedronTest, NonnegativeScalingTest) {
>   HPolyhedron H = HPolyhedron::MakeUnitBox(3);
> ```

I think for better coverage, we should use an H-polyhedron that isn't unit box (namely H = eye and d = ones(3,1)).

**Review by RussTedrake — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/15286)* status: 7 unresolved discussions, LGTM missing from assignee hongkai-dai, needs platform reviewer assigned, needs at least two assigned reviewers (waiting on @hongkai-dai and @TobiaMarcucci)

___
*[geometry/optimization/convex_set.h, line 98 at r2](https://reviewable.io/reviews/robotlocomotion/drake/15286#-MdWeujHyImNgBIqqbD6-r2-98:-MdYJdVc50OwHj8YOnX3:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/c706fa2466b34207d28d96270fe235d27b7dde80/geometry/optimization/convex_set.h#L98)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

BTW, when writing `tS + rec(S)`, would it be more clear to use Minkowski sum ⊕ to denote it is the summation of sets?
</blockquote></details>

Done.
___
*[geometry/optimization/convex_set.h, line 102 at r4](https://reviewable.io/reviews/robotlocomotion/drake/15286#-MdXvKAnAp5eUxpU6ePr:-MdYJe_b8KZkw4x85fB-:b-53etbc) ([raw file](https://github.com/robotlocomotion/drake/blob/ae763485c1d12931f2dc529e6a253c475e556f9e/geometry/optimization/convex_set.h#L102)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

I thought we don't require `S is bounded`?
</blockquote></details>

I guess my sentence was unclear.  I've tried again.
___
*[geometry/optimization/hyperellipsoid.cc, line 67 at r4](https://reviewable.io/reviews/robotlocomotion/drake/15286#-MdXwUmK3fdRTPKYCMN4:-MdYK5K_9vdTT5SnqOkw:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/ae763485c1d12931f2dc529e6a253c475e556f9e/geometry/optimization/hyperellipsoid.cc#L67)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

A_cone * vars + b_cone? x is not mentioned before.
</blockquote></details>

Done.
___
*[geometry/optimization/hyperellipsoid.cc, line 84 at r4](https://reviewable.io/reviews/robotlocomotion/drake/15286#-MdXy1iy9wZ9M_Pr6-5Q:-MdYKPuG5ogSYduLKEuY:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/ae763485c1d12931f2dc529e6a253c475e556f9e/geometry/optimization/hyperellipsoid.cc#L84)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

z = A_cone * [x;t]
</blockquote></details>

Done.
___
*[geometry/optimization/hyperellipsoid.cc, line 90 at r4](https://reviewable.io/reviews/robotlocomotion/drake/15286#-MdXyAghAMQkk_gvtKs7:-MdYKQbP6-dX3dAdMww8:b-89ig50) ([raw file](https://github.com/robotlocomotion/drake/blob/ae763485c1d12931f2dc529e6a253c475e556f9e/geometry/optimization/hyperellipsoid.cc#L90)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

This should be `-A * center_` instead of `center_`? And maybe we need a unit test with non-identity `A` and non-zero `center_` to catch it?
</blockquote></details>

Done.  Yikes.  Thanks!  
___
*[geometry/optimization/test/hpolyhedron_test.cc, line 210 at r4](https://reviewable.io/reviews/robotlocomotion/drake/15286#-MdXyc3Y1omaz8A35Kc2:-MdYO0tu1E8diCp_Kv5N:btztf4w) ([raw file](https://github.com/robotlocomotion/drake/blob/ae763485c1d12931f2dc529e6a253c475e556f9e/geometry/optimization/test/hpolyhedron_test.cc#L210)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

I think for better coverage, we should use an H-polyhedron that isn't unit box (namely H = eye and d = ones(3,1)). 
</blockquote></details>

Done.  I agree.

**Review by hongkai-dai — COMMENTED** [review]
:lgtm: +@ericcousineau-tri for platform review please, thanks!

> Reviewed 5 of 5 files at r5.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/15286#-:-MdYT_Ze6dyxTbZekpEw:bfacy11)* status: 2 unresolved discussions, LGTM missing from assignee EricCousineau-TRI(platform) (waiting on @EricCousineau-TRI and @TobiaMarcucci)

**Review by TobiaMarcucci — COMMENTED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/15286#-:-MdZ5CZv5gYpRTbzmf5U:bnfp4nl)* status: 2 unresolved discussions, LGTM missing from assignee EricCousineau-TRI(platform) (waiting on @EricCousineau-TRI and @TobiaMarcucci)

**Review by TobiaMarcucci — COMMENTED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/15286#-:-MdZ5uRT9ATC7a5GD9tw:bnfp4nl)* status: LGTM missing from assignee EricCousineau-TRI(platform) (waiting on @EricCousineau-TRI)

**Review by EricCousineau-TRI — COMMENTED** [review]
:lgtm: platform with a minor BTW on the paper!

> Reviewed 1 of 5 files at r5.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/15286#-:-Md_-Qqh0mdKVlHy6rvs:bxkzb1g)* status: :shipit: complete! all discussions resolved, LGTM from assignees EricCousineau-TRI(platform),hongkai-dai (waiting on @RussTedrake)

___
*[geometry/optimization/convex_set.h, line 26 at r5](https://reviewable.io/reviews/robotlocomotion/drake/15286#-MdYdpUD1hEFug7KqVtB:-MdYdpUE1e_Pefhc14tm:b-ze81p) ([raw file](https://github.com/robotlocomotion/drake/blob/dd2a172d1fddd596739ffa8e21ee3ee3bfb30b45/geometry/optimization/convex_set.h#L26)):*
BTW Is it possible to reference y'all's paper here?
> Marcucci, Tobia, Jack Umenberger, Pablo A. Parrilo, and Russ Tedrake. “Shortest Paths in Graphs of Convex Sets.” January 27, 2021. http://arxiv.org/abs/2101.11565.

**Review by RussTedrake — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/15286)* status: :shipit: complete! all discussions resolved, LGTM from assignees EricCousineau-TRI(platform),hongkai-dai (waiting on @RussTedrake)

___
*[geometry/optimization/convex_set.h, line 26 at r5](https://reviewable.io/reviews/robotlocomotion/drake/15286#-MdYdpUD1hEFug7KqVtB:-Md_3NMnEO5aPQdBuymf:b-idfuxm) ([raw file](https://github.com/robotlocomotion/drake/blob/dd2a172d1fddd596739ffa8e21ee3ee3bfb30b45/geometry/optimization/convex_set.h#L26)):*
<details><summary><i>Previously, EricCousineau-TRI (Eric Cousineau) wrote…</i></summary><blockquote>

BTW Is it possible to reference y'all's paper here?
> Marcucci, Tobia, Jack Umenberger, Pablo A. Parrilo, and Russ Tedrake. “Shortest Paths in Graphs of Convex Sets.” January 27, 2021. http://arxiv.org/abs/2101.11565.
</blockquote></details>

That paper/algorithm will have it's own .h/.cc file soon. (it's at the end of the train).  This part of the class hierarchy is more much general than that one paper.
